src/three-components/SmoothControls.ts

```diff
@@ -0,0 +1,533 @@
+   * controls instance.
+   */
+  applyConstraints(constraints: SmoothControlsConstraints) {
+    Object.assign(this[$constraints], constraints);
+    // Re-evaluates clamping based on potentially new values for min/max
+    // polar, azimuth and radius:
+    this.setOrbit();
+    // Prevent interpolation in the case that any target spherical values
+    // changed (preserving OrbitalControls behavior):
+    this[$spherical].copy(this[$targetSpherical]);
+  }
+
+  /**
+   * Set the absolute orbital position of the camera relative to the configured
+   * scene origin. The change will be applied over a number of frames depending
+   * on configured acceleration and dampening constraints.
+   *
+   * Returns true if invoking the method will result in the camera changing
+   * position and/or rotation, otherwise false.
+   */
+  setOrbit(
+      targetTheta: number = this[$targetSpherical].theta,
+      targetPhi: number = this[$targetSpherical].phi,
+      targetRadius: number = this[$targetSpherical].radius): boolean {
+    const {
+      minimumAzimuthalAngle,
+      maximumAzimuthalAngle,
+      minimumPolarAngle,
+      maximumPolarAngle,
+      nearOrbitRadius,
+      farOrbitRadius
+    } = this[$constraints];
+
+    const {theta, phi, radius} = this[$targetSpherical];
+
+    const nextTheta =
+        clamp(targetTheta, minimumAzimuthalAngle!, maximumAzimuthalAngle!);
+    const nextPhi = clamp(targetPhi, minimumPolarAngle!, maximumPolarAngle!);
+    const nextRadius = clamp(targetRadius, nearOrbitRadius!, farOrbitRadius!);
+
+    if (nextTheta === theta && nextPhi === phi && nextRadius === radius) {
+      return false;
+    }
+
+    this[$targetSpherical].theta = nextTheta;
+    this[$targetSpherical].phi = nextPhi;
+    this[$targetSpherical].radius = nextRadius;
+    this[$targetSpherical].makeSafe();
+
+    return true;
+  }
+
+  /**
+   * Adjust the orbital position of the camera relative to its current orbital
+   * position.
+   */
+  adjustOrbit(deltaTheta: number, deltaPhi: number, deltaRadius: number):
+      boolean {
+    const {theta, phi, radius} = this[$targetSpherical];
+
+    const targetTheta = theta - deltaTheta;
+    const targetPhi = phi - deltaPhi;
+    const targetRadius = radius + deltaRadius;
+
+    return this.setOrbit(targetTheta, targetPhi, targetRadius);
+  }
+
+  /**
+   * Update controls. In most cases, this will result in the camera
+   * interpolating its position and rotation until it lines up with the
+   * designated target orbital position.
+   */
+  update(_time: number, delta: number) {
+    const deltaTheta = this[$targetSpherical].theta - this[$spherical].theta;
+    const deltaPhi = this[$targetSpherical].phi - this[$spherical].phi;
+    const deltaRadius = this[$targetSpherical].radius - this[$spherical].radius;
+    const distance = vector3.set(deltaTheta, deltaPhi, deltaRadius).length();
+    const frames = delta / FRAME_MILLISECONDS;
```

Comment:
I guess it's a matter of opinion. To me, the arbitrary scale is most clearly expressed as frames. After all, the update function's implementation should represent the change that should take place in a single frame.

It's also a matter of opinion what to do when frames are dropped. We could always advance one real frame per update, if that is desired. But, that means that as device performance drops, time to adjust the orbit shoots up. WDYT?

Reply:
I think it's a matter of "frame" in my mind being bound by the browser rendering cycle (e.g. requestAnimationFrame) rather than being `16.6` -- if `delta` is `66.6`, then `frames` will be `4`, even though there's only been one browser render frame, and `4` is the amount of *potential* frames at 60 FPS since the last update.

The updates should (IMO) definitely use the delta time as to not be a frame-dependent animation, and I don't think that behavior should change! This is more a comment on the variable naming and attempting to explain my confusion with the `frame` naming, although certainly not something that should be bikeshedded too much at the moment

Reply:
What would you say is a good alternative concept here? Tick, perhaps?